### DISCUSSION TITLE
species block, setting time-varying number density

### DISCUSSION THREAD

**Discussion by svoransky** [epochpic/epoch#692] [Q&A] [answered]
Dear contributors,
I'm curious if it is possible to set a time-varying density in species block, or it only takes constant values? And if so, which time should be used: simulation time or laser time (pulse duration, in this case)? I would appreciate any insigths.

**Comment by Status-Mirror** [comment, marked as answer]
Hi @svoransky,

You can have spatially varying densities, but not time-varying. This block only controls the initial density in the simulation.

Once you load the particles, the system evolves according to the fields present, and so the density can change due to particle motion. What are you trying to model by controlling the evolution of the cell densities?

Cheers,
Stuart

**Reply by svoransky** [reply]
Dear Stuart,
Thank you for your reply, it helps a lot. I would like to model electron number density according to time-varying laser field. If number density in the species block controls only the initial density, is there other way to set density to vary with time, or is it not implementable in epoch?

**Reply by Status-Mirror** [reply]
I think there may be some confusion here about what EPOCH does. In a Monte Carlo code like Geant4, you set the target densities, and these remain fixed over the simulation. A PIC code like EPOCH is different.

When you set an initial density, the code loads particles to represent this density. When your time-varying laser field comes in, it accelerates these particles, pushing them around the simulation window. This means that the number of particles in a cell can change during the simulation, and so the code evolves its own number density.

The motion of particles in a laser field, and how the currents they cause affect the incident laser, is quite often what people want to model when they use PIC. The idea of manually moving particles to set a time-varying density slightly defeats the purpose of running a PIC code.

Is there a reason why the default particle motion is not appropriate for your work?

Cheers,
Stuart

**Reply by svoransky** [reply]
My appologies for the late reply. Thank you again for clarification, it took me some time to understand the point you raised. Now, i can aswer that default particle motion set in EPOCH is appropriate, and i have found a way to model the number density i needed without time-dependency. Thank you again for taking your time and explaining this!
